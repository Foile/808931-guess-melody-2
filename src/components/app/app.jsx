import React from "react";
import PropTypes from 'prop-types';
import WelcomeScreen from '../screens/welcome-screen/welcome-screen';
import GenreQuestionScreen from '../screens/genre-question-screen/genre-question-screen';
import ArtistQuestionScreen from '../screens/artist-question-screen/artist-question-screen';
import {connect} from 'react-redux';
import {ActionCreator} from "../../reducer";
import GameMistakes from "../game-mistakes/game-mistakes";


class App extends React.PureComponent {
  _getScreen(question, props) {
    if (question === -1) {
      const {
        gameTime,
        maxMistakes,
        onWelcomeScreenClick
      } = props;

      return <WelcomeScreen
        minutes={gameTime}
        mistakes={maxMistakes}
        onClick={onWelcomeScreenClick}
      />;
    }

    const {questions} = props;
    const currentQuestion = questions[question];
    const {
      onUserAnswer,
      mistakes,
      maxMistakes,
      step
    } = this.props;

    switch (currentQuestion.type) {
      case `genre`: return <GenreQuestionScreen
        step={step}
        question={question}
        onAnswer={(userAnswer) => onUserAnswer(
            userAnswer,
            question,
            mistakes,
            maxMistakes
        )}
      />;

      case `artist`: return <ArtistQuestionScreen
        step={step}
        question={question}
        onAnswer={(userAnswer) => onUserAnswer(
            userAnswer,
            question,
            mistakes,
            maxMistakes
        )}
      />;
    }

    return null;
  }

  constructor(props) {
    super(props);

    this.state = {
      question: -1,
    };
  }

  render() {
    const {
      questions,
      step,
    } = this.props;

    return <section className={`game ${`artist`}`}>
      <header className="game__header">
        <a className="game__back" href="#">
          <span className="visually-hidden">Сыграть ещё раз</span>
          <img className="game__logo" src="img/melody-logo-ginger.png" alt="Угадай мелодию" />
        </a>

        <div className="timer__value">
          <span className="timer__mins">05</span>
          <span className="timer__dots">:</span>
          <span className="timer__secs">00</span>
        </div>
        <GameMistakes mistakes={this.state.mistakes}></GameMistakes>
      </header>

      {this._getScreen(questions[step])}
    </section>;
  }
}

App.propTypes = {
  mistakes: PropTypes.number.isRequired,
  maxMistakes: PropTypes.number.isRequired,
  gameTime: PropTypes.number.isRequired,
  questions: PropTypes.arrayOf(PropTypes.shape({
    type: PropTypes.string.isRequired,
    question: PropTypes.object
  }).isRequired
  ),
  step: PropTypes.number.isRequired,
  onUserAnswer: PropTypes.func.isRequired,
  onWelcomeScreenClick: PropTypes.func.isRequired,
};

const mapStateToProps = (state, ownProps) => Object.assign({}, ownProps, {
  step: state.step,
  mistakes: state.mistakes
});

const mapDispatchToProps = (dispatch) => ({
  onWelcomeScreenClick: () => dispatch(ActionCreator.incrementStep()),
  onUserAnswer: (userAnswer, question, mistakes, maxMistakes) => {
    dispatch(ActionCreator.incrementStep());
    dispatch(ActionCreator.incrementMistake(userAnswer,
        question,
        mistakes,
        maxMistakes));
  }
});


export {App};
export default connect(mapStateToProps, mapDispatchToProps)(App);
