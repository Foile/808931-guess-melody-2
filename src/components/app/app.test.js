import React from 'react';
import renderer from "react-test-renderer";
import {App} from './app';

const questions = [
  {
    "type": `artist`,
    "song": {
      "artist": `Quincas Moreira`,
      "src": `https://es31-server.appspot.com/guess-melody/static/music/Blue_Whale.mp3`
    },
    "answers": [
      {
        "picture": `https://es31-server.appspot.com/guess-melody/static/artist/Quincas_Moreira.jpg`,
        "artist": `Quincas Moreira`
      },
      {
        "picture": `https://es31-server.appspot.com/guess-melody/static/artist/Density_n_Time.jpg`,
        "artist": `Density & Time`
      }
    ]
  }];

it(`App correctly renders after relaunch`, () => {
  const tree = renderer
  .create(<App
    mistakes={0}
    maxMistakes={Infinity}
    gameTime={100}
    questions={questions}
    step={-1}
    onUserAnswer={jest.fn()}
    onWelcomeScreenClick={jest.fn()}
  />)
  .toJSON();
  expect(tree).toMatchSnapshot();
});
