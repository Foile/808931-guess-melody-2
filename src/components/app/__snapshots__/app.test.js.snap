// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App correctly renders after relaunch 1`] = `
<section
  className="welcome"
>
  <div
    className="welcome__logo"
  >
    <img
      alt="Угадай мелодию"
      height="83"
      src="img/melody-logo.png"
      width="186"
    />
  </div>
  <button
    className="welcome__button"
<<<<<<< HEAD
    onClick={[Function]}
=======
    onClick={[MockFunction]}
>>>>>>> 6b0e87f61ed2bd62e8161c0d0c4e59974006b201
  >
    <span
      className="visually-hidden"
    >
      Начать игру
    </span>
  </button>
  <h2
    className="welcome__rules-title"
  >
    Правила игры
  </h2>
  <p
    className="welcome__text"
  >
    Правила просты:
  </p>
  <ul
    className="welcome__rules-list"
  >
    <li>
      За 
       минут нужно ответить на все вопросы.
    </li>
    <li>
      Можно допустить 
       ошибки.
    </li>
  </ul>
  <p
    className="welcome__text"
  >
    Удачи!
  </p>
</section>
`;
